gdx/src/com/badlogic/gdx/graphics/g2d/ArrayTextureSpriteBatch.java

```diff
@@ -0,0 +1,1518 @@
+		invTexHeight = subImageScaleHeight / texture.getHeight();
+
+		// This is our identifier for the textures
+		final int textureHandle = texture.getTextureObjectHandle();
+
+		// First try to see if the texture is already cached
+		for (int i = 0; i < currentTextureLFUSize; i++) {
+
+			// getTextureObjectHandle() just returns an int,
+			// it's fine to call this method instead of caching the value.
+			if (textureHandle == usedTextures[i].getTextureObjectHandle()) {
+
+				// Increase the access counter.
+				usedTexturesLFU[i]++;
+
+				return i;
+			}
+		}
+
+		// If a free texture unit is available we just use it
+		// If not we have to flush and then throw out the least accessed one.
+		if (currentTextureLFUSize < maxTextureLevels) {
+
+			// Put the texture into the next free slot
+			usedTextures[currentTextureLFUSize] = texture;
+
+			// Increase the access counter.
+			usedTexturesLFU[currentTextureLFUSize]++;
+
+			copyTextureIntoArrayTexture(texture, currentTextureLFUSize);
+
+			return currentTextureLFUSize++;
+
+		} else {
+
+			// We have to flush if there is something in the pipeline already,
+			// otherwise the texture index of previously rendered sprites gets invalidated
+			if (idx > 0) {
+				flush();
+			}
+
+			int slot = 0;
+			int slotVal = usedTexturesLFU[0];
+
+			int max = 0;
+			int average = 0;
+
+			// We search for the best candidate for a swap (least accessed) and collect some data
+			for (int i = 0; i < maxTextureLevels; i++) {
+
+				final int val = usedTexturesLFU[i];
+
+				max = Math.max(val, max);
+
+				average += val;
+
+				if (val <= slotVal) {
+					slot = i;
+					slotVal = val;
+				}
+			}
+
+			// The LFU weights will be normalized to the range 0...100
+			final int normalizeRange = 100;
+
+			for (int i = 0; i < maxTextureLevels; i++) {
+				usedTexturesLFU[i] = usedTexturesLFU[i] * normalizeRange / max;
+			}
+
+			average = (average * normalizeRange) / (max * maxTextureLevels);
+
+			// Give the new texture a fair (average) chance of staying.
+			usedTexturesLFU[slot] = average;
+
+			usedTextures[slot] = texture;
+
+			copyTextureIntoArrayTexture(texture, slot);
+
+			// For statistics
+			currentTextureLFUSwaps++;
+
+			return slot;
+		}
+	}
+
+	/** Copies a Texture to the internally managed Array Texture.
+	 * @param texture The Texture to copy onto the Array Texture.
+	 * @param slot The slice of the Array Texture to copy the texture onto. */
+	private void copyTextureIntoArrayTexture (Texture texture, int slot) {
+
+		copyFramebuffer.bind();
+
+		Gdx.gl30.glFramebufferTexture2D(GL30.GL_READ_FRAMEBUFFER, GL30.GL_COLOR_ATTACHMENT0, GL30.GL_TEXTURE_2D,
+			texture.getTextureObjectHandle(), 0);
+
+		Gdx.gl30.glReadBuffer(GL30.GL_COLOR_ATTACHMENT0);
+
+		Gdx.gl30.glBindTexture(GL30.GL_TEXTURE_2D_ARRAY, arrayTextureHandle);
+
+		Gdx.gl30.glCopyTexSubImage3D(GL30.GL_TEXTURE_2D_ARRAY, 0, 0, 0, slot, 0, 0, copyFramebuffer.getWidth(),
+			copyFramebuffer.getHeight());
+
+		FrameBuffer.unbind();

```

Comment:
Won't this break custom framebuffers?

Reply:
Any idea how to navigate around that issue? My idea would be to try and read out the initial state, use the copying framebuffer and then restore the initial state.

Reply:
Yes, this would be the way to handle it (see [here](https://github.com/crykn/libgdx-screenmanager/blob/master/src/main/java/de/eskalon/commons/utils/graphics/NestableFrameBuffer.java#L148) for an example of a nestable framebuffer). Sadly, this gets a littble bit more complicated, as the current GWT implementation doesn't support reading `GL_FRAMEBUFFER_BINDING` (see #4688). There is a workaround by @metaphore for this though, see [here](https://github.com/crashinvaders/gdx-vfx/blob/master/gdx-vfx/gwt/src/com/crashinvaders/vfx/gwt/GwtVfxGlExtension.java#L36-L68).

Reply:
Thanks! Alright the framebuffer restoring was a treat to implement, i don't feel good about adding any GWT specific code to this class whatsoever though. I'll only incorporate this if there is an official LibGDX-esque way to access this GWT-specific routine.

Reply:
This sounds reasonable, but I think you should add a note in the documentation that the batch doesn't work with GWT.

Reply:
Almost forgot about that, i think i'll add a conditional to the FBO save/restore instead so that WebGL can at least use the Batch, albeit without being able to render to an FBO. I'll add some documentation too.